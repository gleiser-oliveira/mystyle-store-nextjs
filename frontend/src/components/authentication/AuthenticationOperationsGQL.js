import gql from 'graphql-tag';
import { perPage } from '../../config';

export const SIGNUP_MUTATION = gql`
    mutation SIGNUP_MUTATION($email: String!, $name: String!, $password: String!) {
        signup(email: $email, name: $name, password: $password) {
            id
            email
            name
        }
    }
`;

export const SIGNIN_MUTATION = gql`
    mutation SIGNIN_MUTATION($email: String!, $password: String!) {
        signin(email: $email, password: $password) {
            id
            email
            name
        }
    }
`;

export const SIGNOUT_MUTATION = gql`
    mutation SIGNOUT_MUTATION {
        signout {
            message
        }
    }
`;

export const CURRENT_USER_QUERY = gql`
    query CURRENT_USER_QUERY { 
        me {
            id
            email
            name
            permissions
            cart {
                id
                quantity
                product {
                    id
                    price
                    title
                    image
                    description
                }
            }
        }
    }
`;

export const REQUEST_PASSWORD_RESET_MUTATION = gql`
    mutation REQUEST_PASSWORD_RESET_MUTATION($email: String!) {
        requestPasswordReset(email: $email) {
            message
        }
    }
`;

export const RESET_PASSWORD_MUTATION = gql`
    mutation RESET_PASSWORD_MUTATION($resetToken: String!, $password: String!, $confirmPassword: String!) {
        resetPassword(resetToken: $resetToken, password: $password, confirmPassword: $confirmPassword) {
            id
            name
            email
        }
    }
`;

export const HAS_PERMISSION_QUERY = gql`
    query { 
        users {
            id
            email
            name
            permissions
        }
    }
`;

export const ALL_USERS_QUERY = gql`
    query ALL_USERS_QUERY($skip: Int = 0, $first: Int = ${perPage}) { 
        users(first: $first, skip: $skip, orderBy: name_ASC) {
            id
            email
            name
            permissions
        }
    }
`;

export const USERS_PAGINATION_QUERY = gql`
    query USERS_PAGINATION_QUERY {
        usersConnection {
            aggregate {
                count
            }
        }
    }
`;

export const ALL_PERMISSIONS_QUERY = gql`
    query { 
        permissions {
          id
          permission
          description
        }
    }
`;

export const UPDATE_USER_PERMISSIONS_MUTATION = gql`
    mutation UPDATE_USER_PERMISSIONS_MUTATION($id: ID! $permissions: [Permission]) {
        updateUserPermissions(id: $id permissions: $permissions) {
            id
            email
            name
            permissions
        }
    }
`;
