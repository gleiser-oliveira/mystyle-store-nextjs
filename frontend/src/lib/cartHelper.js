export const numberOfProducts = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.quantity), 0);
}

export const calcTotalCartPrice = (cart) => {
    return cart.reduce((a, b) => a + (b.product && b.product.price * b.quantity), 0);
}

export const formatPrice = (priceInCents) => {
    const options = {
        style: 'currency',
        currency: 'USD',
        minimumFractionDigits: 2,
    };
    // if its a whole, dollar amount, leave off the .00
    if (priceInCents % 100 === 0)
        options.minimumFractionDigits = 0;

    const formatter = new Intl.NumberFormat('en-US', options);
    return formatter.format(priceInCents / 100);
}