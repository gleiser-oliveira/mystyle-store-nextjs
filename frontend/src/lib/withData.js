import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { LOCAL_STATE_QUERY } from './ClientGQL/LocalGQL';
import { DOCKER_BACKEND_URL, BROWSER_TO_BACKEND_URL } from '../config';

function createClient({ headers }) {
    // if using docker, the URI to the Yoga server must be either:
    // DOCKER_BACKEND_URL = the name of the backend server set in the docker network (used in SSR fetches)
    // BROWSER_TO_BACKEND_URL = the URL of the backend server that can be accessed from the client (used in the client's browser fetches)
    const isInsideFrontendServer = typeof window === 'undefined';
    const uri = isInsideFrontendServer ? DOCKER_BACKEND_URL : BROWSER_TO_BACKEND_URL;
    return new ApolloClient({
        uri: uri,
        request: operation => {
            operation.setContext({
                fetchOptions: {
                    credentials: 'include',
                },
                headers,
            });
        },
        // local data
        clientState: {
            resolvers: {
                Mutation: {
                    toggleCart(_, variables, client) {
                        const { cartOpen } = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        // toggle the state
                        const data = {
                            data: { cartOpen: !cartOpen },
                        };
                        client.cache.writeData(data);
                        return data;
                    },
                },
                Query: {
                    localState(_, variables, client) {
                        const localState = client.cache.readQuery({
                            query: LOCAL_STATE_QUERY,
                        });
                        return localState;
                    }
                }
            },
            defaults: {
                cartOpen: false,
            }
        }
    });
}

// TODO: check getDataFromTree warning - apollo
export default withApollo(createClient);