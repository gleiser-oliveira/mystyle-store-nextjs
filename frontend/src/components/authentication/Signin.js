import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import Link from 'next/link';
import { Form } from '../theme/generalStyles';
import { SIGNIN_MUTATION, CURRENT_USER_QUERY } from './AuthenticationOperationsGQL';
import { ErrorMessage, ConfirmationMessage } from '../message/Message';

class Signin extends Component {
    state = {
        name: '',
        email: '',
        password: '',
    };

    saveToState = (e) => {
        this.setState({ [e.target.name]: e.target.value });
    }

    handleSubmit = async(e, signin) => {
        e.preventDefault();
        const res = await signin();
        this.setState({
            name: '',
            email: '',
            password: '',
            message: `User ${res.data.signin.name} logged successfully!`,
        });
    }

    render() {
        return (
            <Mutation mutation={SIGNIN_MUTATION} variables={{ email: this.state.email, password: this.state.password }} refetchQueries={[{ query: CURRENT_USER_QUERY }]}>
                {(signin, payload) => (
                    <Form method='post' onSubmit={e => { this.handleSubmit(e, signin); }}>
                        <fieldset disabled={payload.loading} aria-busy={payload.loading}>
                            <h2>Sign into your Account</h2>
                            {payload.error ? <ErrorMessage error={payload.error} /> : null}
                            {!payload.error && this.state.message ? <ConfirmationMessage message={this.state.message} /> : null}
                            <label htmlFor='email'>
                                Email
                                <input
                                    type='email'
                                    name='email'
                                    placeholder='email'
                                    value={this.state.email}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <label htmlFor='password'>
                                Password
                                <input
                                    type='password'
                                    name='password'
                                    placeholder='password'
                                    value={this.state.password}
                                    onChange={this.saveToState}
                                />
                            </label>
                            <Link href={{
                                pathname: '/resetPassword',
                                query: { email: this.state.email }
                            }}>
                                <a>Forgot your password? Click here!</a>
                            </Link>
                            <button type='submit'>Sign In!</button>
                        </fieldset>
                    </Form>
                )}
            </Mutation>
        );
    }
}

export default Signin;
