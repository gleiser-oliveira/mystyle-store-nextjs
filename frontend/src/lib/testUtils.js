import casual from 'casual';
import { act } from 'react-dom/test-utils';
import { CURRENT_USER_QUERY } from '../components/authentication/AuthenticationOperationsGQL';

// seeded for constant results
casual.seed(123);

const fakeUser = overrides => ({
    __typename: 'User',
    id: '4234',
    name: casual.name,
    email: casual.email,
    permissions: ['USER', 'PRODUCT_CREATE'],
    cart: [],
    ...overrides,
});

const fakeUser2 = () => ({
    __typename: 'User',
    id: '2222',
    name: casual.name,
    email: casual.email,
    permissions: ['USER'],
    orders: [],
    cart: [],
});

const fakeAdminUser = () => ({
    __typename: 'User',
    id: '1111',
    name: casual.name,
    email: casual.email,
    permissions: ['ADMIN', 'PRODUCT_CREATE'],
    orders: [],
    cart: [],
});

const fakeProduct = () => ({
    __typename: 'Product',
    id: 'QWERTY0123',
    title: 'A Fake Product',
    price: 5000,
    description: 'This is a fake product for testing.',
    image: 'product.jpg',
    largeImage: 'largeProduct.jpg',
    user: fakeUser(),
});

const fakeProduct2 = () => ({
    __typename: 'Product',
    id: 'AS012',
    title: 'Another Fake Product',
    price: 2999,
    description: 'This is another fake product for testing.',
    image: 'product2.jpg',
    largeImage: 'bigProduct2.jpg',
    user: fakeUser(),
});

const fakeOrderProduct = () => ({
    __typename: 'OrderProduct',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
});

const fakeOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    products: [fakeOrderProduct(), fakeOrderProduct()],
    createdAt: 'May 1, 2020 00:00:00',
    user: fakeUser(),
});


const fakeCartProduct = overrides => ({
    __typename: 'CartProduct',
    id: 'prod123',
    quantity: 2,
    product: fakeProduct(),
    user: fakeUser(),
    ...overrides,
});

const fakeCartProduct2 = overrides => ({
    __typename: 'CartProduct',
    id: 'prod312',
    quantity: 2,
    product: fakeProduct2(),
    user: fakeUser(),
    ...overrides,
});

const cartUser = fakeUser();
const fakeCart = () => ([
    fakeCartProduct({user: cartUser}),
    fakeCartProduct2({user: cartUser}),
]);

// Fake LocalStorage
class LocalStorageMock {
    constructor() {
        this.store = {};
    }

    clear() {
        this.store = {};
    }

    getItem(key) {
        return this.store[key] || null;
    }

    setItem(key, value) {
        this.store[key] = value.toString();
    }

    removeItem(key) {
        delete this.store[key];
    }
}

// mocks
const signedInMocks = (user = fakeUser) => { return [
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: user(),
            }
        }
    }
]};

const notSignedInMocks = () => ([
    {
        request: { query: CURRENT_USER_QUERY },
        result: {
            data: {
                me: null,
            }
        }
    }
]);

const updateWrapper = async (wrapper, ms = 0) => {
    await act(async () => {
        await new Promise(resolve => setTimeout(resolve, ms));
        wrapper.update();
    });
}

export {
    LocalStorageMock,
    fakeProduct,
    fakeProduct2,
    fakeUser,
    fakeUser2,
    fakeAdminUser,
    fakeCartProduct,
    fakeCartProduct2,
    fakeOrder,
    fakeCart,
    fakeOrderProduct,
    signedInMocks,
    notSignedInMocks,
    updateWrapper,
};
