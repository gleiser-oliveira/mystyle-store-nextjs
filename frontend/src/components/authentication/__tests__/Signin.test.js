import { mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { ApolloConsumer } from 'react-apollo';
import toJSON from 'enzyme-to-json';
import Signin from '../Signin';
import { SIGNIN_MUTATION, CURRENT_USER_QUERY } from '../AuthenticationOperationsGQL';
import { updateWrapper, fakeUser } from '../../../lib/testUtils';

const usr = fakeUser();

const mocks = [
    {
        request: {
            query: SIGNIN_MUTATION,
            variables: {
                email: usr.email,
                name: usr.name,
                password: 'pass',
            },
        },
        result: {
            data: {
                signin: {
                    __typename: 'User',
                    id: usr.id,
                    email: usr.email,
                    name: usr.name,
                },
            },
        },
    },
    {
        request: {
            query: CURRENT_USER_QUERY,
        },
        result: {
            data: {
                me: {
                    ...usr,
                }
            }
        }
    }
];

describe('testing the <Signin/> component', () => {
    it('renders', () => {
        const wrapper = mount(
            <MockedProvider>
                <Signin/>
            </MockedProvider>
        );

        expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
    });
    it('calls the signin mutation', async() => {
        let apolloClient;
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <ApolloConsumer>
                    {client => {
                        apolloClient = client;
                        return <Signin/>;
                    }}
                </ApolloConsumer>
            </MockedProvider>
        );
        await updateWrapper(wrapper);

        const emailInput = wrapper.find('input[name="email"]');
        const passwordInput = wrapper.find('input[name="password"]');

        emailInput.simulate('change', { target: { name: 'email', value: usr.email } });
        passwordInput.simulate('change', { target: { name: 'password', value: 'pass' } });
        await updateWrapper(wrapper);

        const user = await apolloClient.query({ query: CURRENT_USER_QUERY });

        expect(user.data.me).toMatchObject(usr);
    });
});
