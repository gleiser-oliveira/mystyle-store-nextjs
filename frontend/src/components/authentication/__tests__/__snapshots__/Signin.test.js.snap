// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the <Signin/> component renders 1`] = `
<form
  className="generalStyles__Form-sc-1oz5zu1-1 kCRZkl"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign into your Account
    </h2>
    <label
      htmlFor="email"
    >
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="email"
        type="email"
        value=""
      />
    </label>
    <label
      htmlFor="password"
    >
      Password
      <input
        name="password"
        onChange={[Function]}
        placeholder="password"
        type="password"
        value=""
      />
    </label>
    <Link
      href={
        Object {
          "pathname": "/resetPassword",
          "query": Object {
            "email": "",
          },
        }
      }
    >
      <a
        href="/resetPassword?email="
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        Forgot your password? Click here!
      </a>
    </Link>
    <button
      type="submit"
    >
      Sign In!
    </button>
  </fieldset>
</form>
`;
